src/font.rs

```diff
@@ -114,6 +114,273 @@ impl<'a> Loca<'a> {
+                        }
+                        let glyph_index = glyph_array_value as i16 + id_delta;
+                        Some(glyph_index as u16)
+                    };
+                }
+            } else {
+                size /= 2;
+                index += size;
+            }
+        }
+        None
+    }
+}
+
+impl<'a> Debug for Encoding<'a> {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        f.debug_struct("Encoding")
+            .field("format", &self.get_format())
+            .field("length", &self.get_length())
+            .field("language", &self.get_language())
+            .field("segCountX2", &self.get_seg_count_x_2())
+            .field("searchRange", &self.get_search_range())
+            .field("entrySelector", &self.get_entry_selector())
+            .field("rangeShift", &self.get_range_shift())
+            .field("endCounts", &self.get_end_counts())
+            .field("startCounts", &self.get_start_counts())
+            .field("idDeltas", &self.get_id_deltas())
+            .field("idRangeOffsets", &self.get_id_range_offsets())
+            .finish()
+    }
+}
+
+struct Cmap<'a>(&'a [u8]);
+
+impl<'a> Cmap<'a> {
+    fn get_version(&'a self) -> u16 {
+        let &Cmap(data) = self;
+        get_u16(data, 0).unwrap()
+    }
+
+    fn get_num_tables(&'a self) -> u16 {
+        let &Cmap(data) = self;
+        get_u16(data, 2).unwrap()
+    }
+
+    fn get_encoding_record(&'a self, index: u16) -> Option<EncodingRecord<'a>> {
+        if index > self.get_num_tables() {
+            return None;
+        }
+        let &Cmap(data) = self;
+        let enc_offset = (index * 8 + 4) as usize;
+        let encoding_data = &data[enc_offset as usize..(enc_offset + 12) as usize];
+        Some(EncodingRecord(encoding_data))
+    }
+
+    fn get_encoding_records(&'a self) -> Vec<EncodingRecord> {
+        let mut encodings = vec![];
+        for i in 0..self.get_num_tables() {
+            encodings.push(self.get_encoding_record(i).unwrap());
+        }
+        encodings
+    }
+
+    fn get_encoding(&'a self, index: u16) -> Option<Encoding<'a>> {
+        if index > self.get_num_tables() {
+            return None;
+        }
+        let &Cmap(data) = self;
+        let record = self.get_encoding_record(index).unwrap();
+        let subtable_len = get_u16(data, (record.get_offset() + 2) as usize).unwrap() as u32;
+        let encoding_data =
+            &data[record.get_offset() as usize..(record.get_offset() + subtable_len) as usize];
+        Some(Encoding(encoding_data))
+    }
+
+    fn get_encodings(&'a self) -> Vec<Encoding> {
+        let mut encodings = vec![];
+        for i in 0..self.get_num_tables() {
+            encodings.push(self.get_encoding(i).unwrap());
+        }
+        encodings
+    }
+
+    pub fn lookup_glyph_id(&'a self, code_point: u16) -> Option<u16> {
+        for index in 0..self.get_num_tables() {
+            let encoding = self.get_encoding(index).unwrap();
```

Comment:
This logic is problematic; you're basically taking the first encoding no matter which one it is. Since fonts these days are effectively required to have a Unicode BMP encoding (platform id 3, encoding id 1), you should detect whether that's available. Choosing a "preferred" encoding is something that can happen at cmap creation time, to avoid redoing the search.

Also, if the first encoding is not format 4, it will always return None (ie it's not even checking success).